src/test/java/com/google/cloud/spanner/jdbc/TagMockServerTest.java

```diff
@@ -0,0 +1,209 @@
+import static org.junit.Assert.assertTrue;
+
+import com.google.cloud.spanner.Dialect;
+import com.google.cloud.spanner.MockSpannerServiceImpl.StatementResult;
+import com.google.cloud.spanner.connection.AbstractMockServerTest;
+import com.google.spanner.v1.ExecuteBatchDmlRequest;
+import com.google.spanner.v1.ExecuteSqlRequest;
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import org.junit.After;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+import org.junit.runners.Parameterized.Parameter;
+import org.junit.runners.Parameterized.Parameters;
+
+@RunWith(Parameterized.class)
+public class TagMockServerTest extends AbstractMockServerTest {
+  @Parameter public Dialect dialect;
+
+  @Parameters(name = "dialect = {0}")
+  public static Object[] data() {
+    return Dialect.values();
+  }
+
+  @Before
+  public void setupDialect() {
+    mockSpanner.putStatementResult(StatementResult.detectDialectResult(this.dialect));
+  }
+
+  @After
+  public void clearRequests() {
+    mockSpanner.clearRequests();
+  }
+
+  private String getVariablePrefix() {
+    return dialect == Dialect.POSTGRESQL ? "spanner." : "";
+  }
+
+  private String createUrl() {
+    return String.format(
+        "jdbc:cloudspanner://localhost:%d/projects/%s/instances/%s/databases/%s?usePlainText=true;autoCommit=false",
+        getPort(), "proj", "inst", "db" + (dialect == Dialect.POSTGRESQL ? "pg" : ""));
+  }
+
+  private Connection createConnection() throws SQLException {
+    return DriverManager.getConnection(createUrl());
+  }
+
+  @Test
+  public void testStatementTagForQuery() throws SQLException {
+    try (Connection connection = createConnection()) {
+      connection
+          .createStatement()
+          .execute(String.format("set %sstatement_tag='my-tag'", getVariablePrefix()));
+      try (ResultSet resultSet =
+          connection.createStatement().executeQuery(SELECT_RANDOM_STATEMENT.getSql())) {
+        assertTrue(resultSet.next());
+      }
+      assertEquals(1, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+      ExecuteSqlRequest request = mockSpanner.getRequestsOfType(ExecuteSqlRequest.class).get(0);
+      assertEquals("my-tag", request.getRequestOptions().getRequestTag());
+
+      // Verify that the tag is cleared after having been used.
+      mockSpanner.clearRequests();
+      try (ResultSet resultSet =
+          connection.createStatement().executeQuery(SELECT_RANDOM_STATEMENT.getSql())) {
+        assertTrue(resultSet.next());
+      }
+      assertEquals(1, mockSpanner.countRequestsOfType(ExecuteSqlRequest.class));
+      request = mockSpanner.getRequestsOfType(ExecuteSqlRequest.class).get(0);
+      assertEquals("", request.getRequestOptions().getRequestTag());
+    }
+  }
+
+  @Test
+  public void testTransactionTagForQuery() throws SQLException {
```

Comment:
Cant we we combine the above tests with this as the setup is same just the assertion is different?